Sources/NIOCore/ByteBuffer-QUICLengthPrefix.swift

```diff
@@ -0,0 +1,171 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if os(Windows)
+import ucrt
+#elseif canImport(Darwin)
+import Darwin
+#elseif canImport(Glibc)
+import Glibc
+#elseif canImport(Musl)
+import Musl
+#elseif canImport(Bionic)
+import Bionic
+#else
+#error("The Byte Buffer module was unable to identify your C library.")
+#endif
+
+extension ByteBuffer {
+    /// Read a QUIC variable-length integer, moving the `readerIndex` appropriately
+    /// If there are not enough bytes, nil is returned
+    ///
+    /// [RFC 9000 § 16](https://www.rfc-editor.org/rfc/rfc9000.html#section-16)
+    public mutating func readQUICVariableLengthInteger() -> Int? {
+        guard let firstByte = self.getInteger(at: self.readerIndex, as: UInt8.self) else {
+            return nil
+        }
+
+        // Look at the first two bits to work out the length, then read that, mask off the top two bits, and
+        // extend to integer.
+        switch firstByte & 0xC0 {
+        case 0x00:
+            // Easy case.
+            self.moveReaderIndex(forwardBy: 1)
+            return Int(firstByte & ~0xC0)
+        case 0x40:
+            // Length is two bytes long, read the next one.
+            return self.readInteger(as: UInt16.self).map { Int($0 & ~(0xC0 << 8)) }
+        case 0x80:
+            // Length is 4 bytes long.
+            return self.readInteger(as: UInt32.self).map { Int($0 & ~(0xC0 << 24)) }
+        case 0xC0:
+            // Length is 8 bytes long.
+            return self.readInteger(as: UInt64.self).map { Int($0 & ~(0xC0 << 56)) }
+        default:
+            fatalError("Unreachable")
+        }
+    }
+
+    /// Write a QUIC variable-length integer.
+    ///
+    /// [RFC 9000 § 16](https://www.rfc-editor.org/rfc/rfc9000.html#section-16)
+    @discardableResult
+    public mutating func writeQUICVariableLengthInteger(_ value: Int) -> Int {
```

Comment:
Any reason this shouldn't be generic over `FixedWidthInteger`? It would avoid the awkward need to include UInt in our API surface.